src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java

```diff
@@ -1891,7 +1891,8 @@ public void visitCtCircularTypeReference(CtCircularTypeReference reference) {
 
 	private <T> boolean hasDeclaringTypeWithGenerics(CtTypeReference<T> reference) {
 		// If current reference use generic types, we don't need this hack.
-		if (reference.getActualTypeArguments().size() != 0) {
+		if (reference.getActualTypeArguments().size() != 0
+				&& reference.getDeclaringType().getActualTypeArguments().size() == 0) {
```

Comment:
Your additional condition. Potentially, `reference.getDeclaringType()` can be null.